### ISSUE TITLE
UDP connections may be incorrectly dropped because of former DNS redirect policy

### ISSUE THREAD

**Issue by pchaigno** [cilium/cilium#17459]
If we used to have a policy redirecting traffic to a UDP proxy (typically the DNS proxy), after this policy is removed new connections may still be redirected to the proxy because they match an old, redirected connection in the conntrack map. As a result packets are dropped by the userspace proxy.

This incorrect redirection to the proxy can happen for up to 6 minutes after the DNS/L7 policy was removed in default settings. Every 5 minutes, we garbage collect UDP connections that didn't have activity for 1 minutes.

We may need to implement a grace period in the proxy where packets are not dropped to avoid this.

See https://github.com/cilium/cilium-cli/issues/367#issuecomment-920218323 and https://github.com/cilium/cilium-cli/issues/367#issuecomment-922937121 for details.

**Comment by joestringer** [comment]
Minor question, where does the `Every 5 minutes` come from? Conntrack GC is typically dynamic and may take much longer. Is there another mechanism that is ensuring that this will be analyzed within 5 minutes?

If there is a grace period for disabling the proxy, we will still need to ensure that the proxy is correctly implementing the new forwarding policy during that time, for instance if the policy changes to simply drop the traffic, the proxy would need to also follow that policy. And presumably that grace period should be tied into the conntrack GC interval (for instance, by triggering a new GC iteration, then after GC runs, re-evaluating whether the proxy can disable its logic)

Alternative options are possibly more invasive, eg the datapath could consult the policymap again for the original direction flow to understand whether the current policy is to forward the traffic towards the proxy, rather than relying on conntrack.

**Comment by pchaigno** [comment]
> Minor question, where does the `Every 5 minutes` come from? Conntrack GC is typically dynamic and may take much longer. Is there another mechanism that is ensuring that this will be analyzed within 5 minutes?

No, you're correct. I didn't measure this, just checked the code to see what I could expect for a default deployment. So I guess it's more of a: if you want to *maximize* chances of reproducing, you should ensure the two connections are within 6 minutes of each other.

> If there is a grace period for disabling the proxy, we will still need to ensure that the proxy is correctly implementing the new forwarding policy during that time, for instance if the policy changes to simply drop the traffic, the proxy would need to also follow that policy.

Yep, I'm not sure how feasible this grace period idea is. I'm a bit out of depth here (w.r.t. proxy internals). We should probably set up an ad-hoc call to discuss this at some point.

> And presumably that grace period should be tied into the conntrack GC interval (for instance, by triggering a new GC iteration, then after GC runs, re-evaluating whether the proxy can disable its logic)

Absolutely, I forgot to mention that. Thanks!

**Comment by christarazi** [comment]
Are you referring to https://github.com/cilium/cilium/blob/1d6a649c2b93f8dce62fbd127a7458c6cc50db0e/pkg/proxy/proxy.go#L690 or something else?

**Comment by pchaigno** [comment]
No idea, it's been years.

**Comment by mathpl** [comment]
@mhofstetter once you get to this one have a chat with @jrajahalme to get you on the right track he should have a good idea of what's happening here. Thanks!

**Comment by pchaigno** [comment]
@brb Couldn't the same kind of CT reuse bug cause us to use the wrong source IP for SNAT?

I have a case where, after X successful requests to a NodePort, we suddenly fail because we pick the wrong IP for SNAT (CiliumInternalIP instead of NodeInternalIP). I know the same workflow is currently affected by the present issue because we also have a failure where we see a redirect to the proxy due to an incorrect CT entry reuse. I'm wondering if the two types of failures could have the same root cause.

**Comment by brb** [comment]
> Couldn't the same kind of CT reuse bug cause us to use the wrong source IP for SNAT?

It's totally plausible. I find that usually Hubble flows + CT and SNAT maps content is enough to determine that.

**Comment by joestringer** [comment]
@mhofstetter @mathpl were you still planning to take a look at this? If not, we can remove you from assignee to make it more clear that it's open for others to pick up.

**Comment by mathpl** [comment]
:+1:, wish me to find a have a new assignee or we do we have a willing party here?

**Comment by mathpl** [comment]
^ missclick, sorry!

**Comment by julianwiedmann** [comment]
> If we used to have a policy redirecting traffic to a UDP proxy (typically the DNS proxy), after this policy is removed new connections may still be redirected to the proxy because they match an old, redirected connection in the conntrack map.

I believe we substantially improved this situation with https://github.com/cilium/cilium/pull/33222. CT entries are now re-created if their `proxy_redirect` flag doesn't match the result from the network policy lookup.

Thus the described situation would only occur for a connection where we see **only** reply traffic (since we don't perform a network policy lookup for such packets). The first packet in forward direction should fix the problem.
